platform/m6/sub/101a/boot.c

```diff
@@ -0,0 +1,248 @@
+#include "cache.h"
+
+#include "../../../generic/mmu_utils.h"
+#include "boot.h"
+#include "debug.h"
+
+#define CHDK_MMUTBL_START 0xdffda400
+#define CHDK_PAGES_START  0x11d0000
+#define CANON_TABLES_AT   0xdffc0000
+
+static int no_chdk_please = 0;
+
+void CreateTask_spytask() {
+
+}
+
+void startup_mode_fix(void) {
+	int startupmode = *(int*) 0x9eb0;
+	switch (startupmode) {
+	case 0x1000000: // rec
+		break;
+	default:
+		no_chdk_please = 1;
+	}
+}
+
+void __attribute__((naked,noinline,aligned(4))) patch_E0020668() {
+	asm volatile (
+			"bl         CreateTask_spytask\n"
+			"bl         sub_E02796AE\n" // execute overwritten instruction
+			"bl         sub_E010E954\n"// execute overwritten instruction
+			"ldr        pc, =0xe0020671\n"// -> back to ROM
+	);
+}
+
+void plant_hacks_for_e0020000(unsigned addr) {
+	// switch off card instead of exec'ing diskboot
+	*(unsigned*) (addr + 0x0644) = BL_INSTR_T1(0xE0020644, 0xE04B3F3A);
+	// conditionally start SpyTask
+	place_fw_patch_t2_64b(patch_E0020668, addr + 0x0668);
+}
+
+void __attribute__((naked,noinline,aligned(4))) patch_E0079DA4() {
+	asm volatile (
+			"bl         sub_e007006e\n" // startup mode inquiry
+			"bl         startup_mode_fix\n"// +
+			"bl         sub_e007a49c\n"// <kbd_read_keys>
+			"ldr        r3, =0x6a1a8\n"
+			"ldr        pc, =0xe0079dad\n"// -> back to ROM
+	);
+}
+
+void plant_hacks_for_e0070000(unsigned addr) {
+	// insert startup mode handling
+	place_fw_patch_t2_64b(patch_E0079DA4, addr + 0x9DA4);
+	// nop (benign 16-bit instruction, added to replace 1st half of destroyed 32-bit instruction)
+	*(unsigned short*) (addr + 0x9DA2) = 0xbf00;
+}
+
+// following should be integrated in dcache_clean_all
+void l2_cache_sync(void) {
+	*(int*) 0xc1100730 = 0;
+}
+
+void setup_mmu_for_chdk(int core) {
+	if (core == 0) {
+		unsigned l2adr = CHDK_MMUTBL_START;
+		unsigned rompagecopy = CHDK_PAGES_START;
+
+		// at the beginning, ROM tables are in use
+		// first, make a fresh copy of the ROM translation tables to its usual place in RAM
+		_make_copy_of_canon_mmutables(CANON_TABLES_AT);
+		// get flags of ROM area
+		unsigned flags = get_l2_largepage_flags_from_l1_section(ROMBASEADDR,
+		CANON_TABLES_AT);
+
+		// split first 16MB of ROM area into sections
+		split_l1_supersection(ROMBASEADDR, CANON_TABLES_AT);
+
+		{
+			// create L2 table for following 1MB ROM section
+			replace_section_with_l2_tbl(ROMBASEADDR, CANON_TABLES_AT, l2adr,
+					flags);
+
+			replace_rom_page(ROMBASEADDR + 0x20000, rompagecopy, l2adr, flags);
+			plant_hacks_for_e0020000(rompagecopy);
+			rompagecopy += PAGE_SIZE;
+
+			replace_rom_page(ROMBASEADDR + 0x70000, rompagecopy, l2adr, flags);
+			plant_hacks_for_e0070000(rompagecopy);
+			rompagecopy += PAGE_SIZE;
+
+			l2adr += 0x400;
```

Comment:
why do we need this ?

Reply:
rompagecopy specifies the RAM address where a 64kB ROM page is copied. When all patches are placed in that page, rompagecopy is incremented to point to the next such copied page (if any). l2adr points to the L2 table describing memory translation for a 1 MB section of memory. It's incremented when the current section is done.
This could probably be implemented with macros or functions, but I left it like this "for now".
Note that nothing protects from overusing RAM available for either purpose, the coder needs to take care not to exceed available space..